Add documentation about the user directory search algorithm (#16320)
matrix-org/synapse: docs/user_directory.md

# User Directory API Implementation

The user directory is maintained based on the 'visible' users of a homeserver -
i.e. ones which are local to the server, ones which a local user shares a room

> with, or ones who are present in a publicly viewable room present on the server.

Reviewer: Isn't that last one redundant? 

I think this is getting confused between which users the HS will index and which matches it will actually return to the user?

Author: I think you're right it is mixing concerns.

> The user directory is maintained based on the 'visible' users of a homeserver -
i.e. ones which are local to the server, and ones which a local user shares a room
with.

I think that's accurate? They don't need to be in public rooms.

Reviewer: Yup, I think so.

  - Indexed on the full text search data. Indexed on users.

* `user_directory_stream_pos`. When the initial background update to populate
  the directory is complete, we record a stream position here. This indicates
  that synapse should now listen for room changes and incrementally update
  the directory where necessary.

* `users_in_public_rooms`. Contains associations between users and the public
  rooms they're in.  Used to determine which users are in public rooms and should
  be publicly visible in the directory.

* `users_who_share_private_rooms`. Rows are triples `(L, M, room id)` where `L`
   is a local user and `M` is a local or remote user. `L` and `M` should be
   different, but this isn't enforced by a constraint.

   Note that if two local users share a room then there will be two entries:
   `(user1, user2, !room_id)` and `(user2, user1, !room_id)`.

## Configuration options

The exact way user search works can be tweaked via some server-level
[configuration options](usage/configuration/config_documentation.md#user_directory).

The information is not repeated here, but the options are mentioned below.

## Search algorithm

> If `search_all_users` is `false`, then results are limited to users who:

Reviewer 2: Can we have a sentence to say what happens if `search_all_users` is True?

1. Are found in the `users_in_public_rooms` table, or
2. Are found in the `users_who_share_private_rooms` where `L` is the requesting
   user and `M` is the search result.

By default, locked users are not returned. If `show_locked_users` is `true` then

> no filtering on the locked status of a user is done.

Author: Yes...well I'm just documenting what we do. 😢

The user provided search term is lowercased and normalized using [NFKC](https://en.wikipedia.org/wiki/Unicode_equivalence#Normalization),
this treats the string as case-insensitive, canonicalizes different forms of the
same text, and maps some "roughly equivalent" characters together.

The search term is then split into words:

* If [ICU](https://en.wikipedia.org/wiki/International_Components_for_Unicode) is

>   available, then the default locale will be used to break the search term into words.

Reviewer 2: To check: is this the system default locale? Or more precisely, whatever the `LC_*` env vars are set to in Synapse's environment?

Author: It uses [`Locale.getDefault()` from pyicu](https://gitlab.pyicu.org/main/pyicu/-/blob/65b471889b14c4eb08f86c3bb8190b8aae6f2326/locale.cpp#L1129-1132), which seems to fall into [this code from ICU](https://github.com/unicode-org/icu/blob/9bce52f003496527b2e37e31ae08521cd35212ac/icu4c/source/common/locid.cpp#L2041-L2052).

I think that corresponds to [this documentation](https://unicode-org.github.io/icu/userguide/locale/#default-locales):

> The default locale is set to be the system locale on that platform.

  (See the [installation instructions](setup/installation.md) for how to install ICU.)
* If unavailable, then runs of ASCII characters, numbers, underscores, and hypens
  are considered words.

### PostgreSQL

The above words are then transformed into two queries:

1. "exact" which matches the parsed words exactly (using [`to_tsquery`](https://www.postgresql.org/docs/current/textsearch-controls.html#TEXTSEARCH-PARSING-QUERIES));
2. "prefix" which matches the parsed words as prefixes (using `to_tsquery`).

> Results are sorted by a rank derived by:

Reviewer 2: - Found it a bit hard to understand what these queries search for.
- Suggest avoiding `rank` here to avoid confusion with the SQL `rank` stuff below.
- Found it easier to understand the multipliers are starting from a score of 1

```suggestion
We find all rows in the `user_directory_search` table whose usernames match one (or both) of these queries.
The results are sorted by a score.
The initial score is 1, which is then multiplied by various conditional factors. These factors are:
```

> * 4x if a user ID exists.

Reviewer 2: Err, when would this _not_ be the case?

Author: I had the same question. I think it is not-nullable.

Author: ```sql
=> \d user_directory_search
        Table "user_directory_search"
 Column  |   Type   | Collation | Nullable | Default 
---------+----------+-----------+----------+---------
 user_id | text     |           | not null | 
 vector  | tsvector |           |          | 
Indexes:
    "user_directory_search_fts_idx" gin (vector)
    "user_directory_search_user_idx" UNIQUE, btree (user_id)
```

Author: I think this actually means the initial weight is 4, not 1? 🤷

Reviewer 2: Effectively, though it won't make much difference: sorting by `val` and `4 * val` should give the same results[^1]!

[^1]: except that the former is less prone to overflow, I _guess_

* 1.2x if the user has a display name set.

> * 1.2x if the user has an avatar set.

Reviewer 2: I guess this is to bias the search towards more "active" users?

Author: "active" or "real" users, I think.

* 3x by the full text search results using the [`ts_rank_cd` function](https://www.postgresql.org/docs/current/textsearch-controls.html#TEXTSEARCH-RANKING)

>   against the "exact" search query; this has four variables with the following weightings:

Reviewer 2: Any idea what values this function produces? My first guess was 0--1, but I'm not sure that's correct. We use a normalisation value of `8` which

> divides the rank by the number of unique words in document

but it's not clear to me what that actually means.

  * `D`: 0.1 for the user ID's domain
  * `C`: 0.1 for unused
  * `B`: 0.9 for the user's display name (or an empty string if it is not set)
  * `A`: 0.1 for the user ID's localpart
* 1x by the full text search results using the `ts_rank_cd` function against the
  "prefix" search query. (Using the same weightings as above.)
* If `prefer_local_users` is `true`, then 2x if the user is local to the homeserver.

### SQLite

> Results are sorted as follows:
> 
> * By the [`rank`](https://www.sqlite.org/windowfunctions.html#built_in_window_functions)
>   of the full text search results using the [`matchinfo` function](https://www.sqlite.org/fts3.html#matchinfo).
> * If `prefer_local_users` is `true`, then if the user is local to the homeserver.
> * If the user has a display name set.
> * If the user has an avatar set.

Reviewer 2: Took me a moment to realise that this is an `ORDER BY a, b, c, d`, whereas postgres is `ORDER BY a * b * c * d`.

Does this help clarify at all?

```suggestion
Results are sorted as by the [`rank`](https://www.sqlite.org/windowfunctions.html#built_in_window_functions)
of the full text search results using the [`matchinfo` function](https://www.sqlite.org/fts3.html#matchinfo).
Equally-ranked results are tiebroken as follows:

* When `prefer_local_users` is `true`, local users are prioritised over remote users.
* Next, users with a display name set are prioritised over those without a displayname.
* Finally, users with an avatar set are prioritised over users without avatars.
```

# User Directory API Implementation

> The user directory is maintained based on the 'visible' users of a homeserver -

Reviewer 2: "visible users of a homeserver" sounds a bit like we control these users, even the remote ones. Maybe:

```suggestion
The user directory is maintained based on users that are 'visible' to the homeserver -
```

room with.

The directory info is stored in various tables, which can sometimes get out of
sync (although this is considered a bug). If this happens, for now the
solution to fix it is to use the [admin API](usage/administration/admin_api/background_updates.md#run)
and execute the job `regenerate_directory`. This should then start a background task to
flush the current tables and regenerate the directory. Depending on the size
of your homeserver (number of users and rooms) this can take a while.

## Data model

There are five relevant tables that collectively form the "user directory".
Three of them track a list of all known users. The last two (collectively called
the "search tables") track which users are visible to each other.

From all of these tables we exclude three types of local user:

- support users
- appservice users
- deactivated users

A description of each table follows:

* `user_directory`. This contains the user ID, display name and avatar of each user.
  - Because there is only one directory entry per user, it is important that it
    only contain publicly visible information. Otherwise, this will leak the
    nickname or avatar used in a private room.

>   - Indexed on rooms. Indexed on users.

Reviewer 2: Aside: I have a strong suspicion that this table is redundant, and could be replaced by some combination of `room_memberships`, `profiles` and maybe `current_state_events`. But we're not fixing that here.

Author: `profiles` is only useful for local users though. And `room_memberships` / `current_state_events` might have room-specific display name / avatars; so this stores each both local and remote user's *global* profile information only.

Reviewer 2: Ahhh yes. It's all coming back to me now!

* `user_directory_search`. To be joined to `user_directory`. It contains an extra
  column that enables full text search based on user IDs and display names.
  Different schemas for SQLite and Postgres are used.

>   - Indexed on the full text search data. Indexed on users.

Reviewer 2: Aside: I always found it strange that this is a separate table. I guess it's to limit the sqlite vs postgres differences to a small table rather than a larger one?

Author: I think it is because you enable "full text search" on a table, but the database engines then back that by multiple tables that are hidden from you.

* `user_directory_stream_pos`. When the initial background update to populate
  the directory is complete, we record a stream position here. This indicates
  that synapse should now listen for room changes and incrementally update

>   the directory where necessary.

Reviewer 2: Could cross-reference to https://matrix-org.github.io/synapse/latest/development/synapse_architecture/streams.html here maybe

  will be used to break the search term into words. (See the
  [installation instructions](setup/installation.md) for how to install ICU.)
* If unavailable, then runs of ASCII characters, numbers, underscores, and hypens
  are considered words.

The queries for PostgreSQL and SQLite are detailed below, by their overall goal
is to find matching users, preferring users who are "real" (e.g. not bots,
not deactivated). It is assumed that real users will have an display name and
avatar set.

### PostgreSQL

The above words are then transformed into two queries:

1. "exact" which matches the parsed words exactly (using [`to_tsquery`](https://www.postgresql.org/docs/current/textsearch-controls.html#TEXTSEARCH-PARSING-QUERIES));
2. "prefix" which matches the parsed words as prefixes (using `to_tsquery`).

Results are composed of all rows in the `user_directory_search` table whose information
matches one (or both) of these queries. Results are ordered by calculating a weighted
score for each result, higher scores are returned first:

* 4x if a user ID exists.
* 1.2x if the user has a display name set.
* 1.2x if the user has an avatar set.

> * 3x by the full text search results using the [`ts_rank_cd` function](https://www.postgresql.org/docs/current/textsearch-controls.html#TEXTSEARCH-RANKING)

Reviewer 2: Could maybe say 

> 0-3x, depending on the full text search result using...

and "0-1x" below?
